### PR TITLE
for #950, update babel to v7

### PR DESCRIPTION
**Pull request by catarak**
I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] is from a uniquely-named feature branch and has been rebased on top of the latest master. (If I was asked to make more changes, I have made sure to rebase onto master then too)
* [x] is descriptively named and links to an issue number, i.e. `Fixes #123`

I upgraded babel to v7, but since I changed a lot of core things I think this needs a review!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,53 +38,69 @@
     "url": "git+https://github.com/catarak/p5.js-web-editor.git"
   },
   "devDependencies": {
-    "babel-eslint": "^7.1.1",
-    "babel-loader": "^7.1.5",
-    "babel-plugin-transform-react-constant-elements": "^6.23.0",
-    "babel-plugin-transform-react-inline-elements": "^6.22.0",
+    "@babel/core": "^7.0.0",
```
Comment (andrewn) [inline package.json]: This is listed in `devDependencies` and `dependencies`.

File: package.json [hunk 2]
```diff
@@ -38,53 +38,69 @@
...
     "webpack-manifest-plugin": "^2.0.4",
-    "react-test-renderer": "^16.6.0",
     "webpack-node-externals": "^1.7.2"
   },
   "engines": {
     "node": "10.15.0",
     "npm": "6.4.1"
   },
   "dependencies": {
+    "@babel/core": "^7.0.0",
```
Comment (andrewn) [inline package.json]: This is listed in `devDependencies` and `dependencies`.

File: webpack/config.dev.js [hunk 3]
```diff
@@ -122,17 +122,8 @@ module.exports = [{
...
-            'stage-0',
-          ],
-          plugins: [
-            [
-              'babel-plugin-webpack-loaders', {
-                'config': path.resolve(__dirname, './config.babel.js'),
-                'verbose': false
-              }
-            ]
+            '@babel/preset-react',
```
Comment (andrewn) [inline webpack/config.dev.js]: Is it possible for the webpack babel-loader plugin to read from `.babelrc` so that `@babel/preset-react` and `@babel/preset-env` don't need to be duplicated in the webpack configs too?

The [documentation](https://github.com/babel/babel-loader) isn't very clear about this.

File: .babelrc [hunk 4]
```diff
@@ -1,13 +1,73 @@
...
+    "@babel/preset-env"
+  ],
   "env": {
     "production": {
       "plugins": [
         "transform-react-remove-prop-types",
-        "transform-react-constant-elements",
-        "transform-react-inline-elements"
+        "@babel/plugin-transform-react-constant-elements",
+        "@babel/plugin-transform-react-inline-elements",
```
Comment (andrewn) [inline .babelrc]: The [`@babel/plugin-transform-react-inline-elements` documentation](https://babeljs.io/docs/en/babel-plugin-transform-react-inline-elements/) is very cryptic but seems to say that for legacy browser support we should also be using the `useBuiltIns` option of `@babel/preset-env`?

> When used alongside @babel/plugin-transform-runtime, polyfills (by default including Symbol) are specifically scoped to not pollute the global scope. This breaks usage with React, as it won't have access to that polyfill and will cause your application to fail in legacy browsers.
> 
> Even if ['@babel/plugin-transform-runtime', { helpers: true, polyfill: false }] is specified, it might still break, since helpers come precompiled.
> 
> In this case, we recommend importing/requiring @babel/polyfill in the entry point of your application and using @babel/preset-env with the useBuiltIns option to only include the polyfills your targets need. Alternatively, you can also import/require core-js/modules/es6.symbol by itself.

File: .babelrc [hunk 5]
```diff
@@ -1,13 +1,73 @@
...
+        "@babel/plugin-syntax-dynamic-import",
```
Comment (andrewn) [inline .babelrc]: The [`@babel/plugin-syntax-dynamic-import` documentation](https://babeljs.io/docs/en/babel-plugin-syntax-dynamic-import#working-with-webpack-and-babel-preset-env) says:

> Currently, @babel/preset-env is unaware that using import() with Webpack relies on Promise internally. Environments which do not have builtin support for Promise, like Internet Explorer, will require both the promise and iterator polyfills be added manually.

File: .babelrc [hunk 6]
```diff
@@ -1,13 +1,73 @@
...
+        "@babel/plugin-syntax-import-meta",
+        "@babel/plugin-proposal-class-properties",
+        "@babel/plugin-proposal-json-strings",
+        [
+          "@babel/plugin-proposal-decorators",
```
Comment (andrewn) [inline .babelrc]: [`@babel/plugin-proposal-decorators`](https://babeljs.io/docs/en/babel-plugin-proposal-decorators#legacy):

> Use the legacy (stage 1) decorators syntax and behavior.
> NOTE: Compatibility with @babel/plugin-proposal-class-properties
> 
> If you are including your plugins manually and using @babel/plugin-proposal-class-properties, make sure that @babel/plugin-proposal-decorators comes before @babel/plugin-proposal-class-properties.
> 
> When using the legacy: true mode, @babel/plugin-proposal-class-properties must be used in loose mode to support the @babel/plugin-proposal-decorators.

**Comment by andrewn** [comment]
Wow, thanks for tackling this!

I've added some comments above based on each plugin's documentation.

### Manual test plan

I've also done the following manual testing:

- Run `npm ci` to ensure clean install of modules
- Run `npm start` to test dev mode

<details>
  <summary>I performed the following manual actions</summary>

- Open site
- Log in
- Edit sketch
- Turn on Auto refresh
- Play/Stop sketch
- Edit index.html
- Edit styles.css
- Rename styles.css -> index.css
- Add folder `folder`
- Add file 'something.js' to folder with `console.log('added');`
- Include `folder/something.js` to `index.html`
- File -> Examples
- Open `Hello P5: Weather`
- Open Settings
- Change font size
- Change theme to Dark
- Enable "Accessible text based canvas" -> Sound
- Open Keyboard Shortcuts modal
- Comment line via keyboard shortcut (Cmd + /)
- Find "position" in editor (Cmd + F)
- File -> Download (check sketch zip opens)
- Sketch -> Add File (test file upload)
- Display in sketch via index.html 
</details>

### Browsers:

- Firefox 66.0.5 (64-bit) macOS Mojave 10.14.4
- Safari 12.1 (14607.1.40.1.4) macOS Mojave 10.14.4

Shall I also test in Microsoft Edge? What versions does the Editor support?

File: package.json (hunk 1 above)
Comment (catarak) [reply package.json]: weird! i used `npm babel-upgrade` to move all of the packages, which must have saved it in both. i'll move it to just `dependencies`

File: .babelrc (hunk 4 above)
Comment (catarak) [reply .babelrc]: upon a slow reading of the documentation, i think it's saying that it will break in legacy browsers only if the project is also using `@babel/plugin-transform-runtime`?

File: .babelrc (hunk 5 above)
Comment (catarak) [reply .babelrc]: i think this mostly affects IE support. to get this to work, does `core-js` need to be added as a dependency? i fell into a hole reading about `@babel/preset-env` with [`useBuiltIns: "usage"`](https://babeljs.io/docs/en/babel-preset-env#usebuiltins-usage-experimental), but can't figure out whether using this feature affects this plugin.

**Comment by catarak** [comment]
> Shall I also test in Microsoft Edge? What versions does the Editor support?

There have been some issues with Edge that I've never been able to figure out/fix (see #27), but as far as I know, Edge should be supported. I have a free BrowserStack plan because this project is open source, should I invite you?

File: .babelrc (hunk 4 above)
Comment (andrewn) [reply .babelrc]: You're right. That seems to be it. And we're not using `@babel/plugin-transform-runtime` so we're all good.

File: .babelrc (hunk 5 above)
Comment (andrewn) [reply .babelrc]: Yes, I think setting `useBuiltIns` to anything other than `false` will insert `core-js` imports for the required polyfills.

`core-js` seems to already be included in `node_modules` because of the following dependencies:

  `@babel/preset-env` ▶️  `core-js-compat` ▶️ `core-js`

**Comment by andrewn** [comment]
> I have a free BrowserStack plan because this project is open source, should I invite you?

@catarak Yes please!

File: .babelrc (hunk 5 above)
Comment (catarak) [reply .babelrc]: okay, i just pushed a change to the webpack configs to support this.

**Comment by catarak** [comment]
when i ran the command `npx babel-upgrade`, it only upgraded babel to 7.0.0—do you think i should try to update it to 7.4?

**Comment by andrewn** [comment]
I think if we're going to all the effort of testing the changes, it's worth upgrading to the latest released version. 🚀

**Comment by catarak** [comment]
cool, i'm going to upgrade everything to LTS

**Comment by catarak** [comment]
okay great. i'm going to merge this!
